src/storage.rs

```diff
@@ -115,32 +136,58 @@ impl<T: Clone + 'static> RingBufferStorage<T> {
         } else {
             self.written - reader_id.written
         };
+
+        let read_index = reader_id.read_index;
+        reader_id.read_index = self.write_index;
+        reader_id.written = self.written;
+
         if num_written > self.max_size {
-            let mut d = self.data
-                .get(self.write_index..self.max_size)
-                .unwrap()
-                .to_vec();
-            d.extend(self.data.get(0..self.write_index).unwrap().to_vec());
-            reader_id.read_index = self.write_index;
-            reader_id.written = self.written;
-            return Err(RBError::LostData(d, num_written - self.max_size));
+            Err(RBError::LostData(
```

Comment:
Ah. I see now. It's the same iterator as in `Ok` but with flag set that you lost some data.
It can be error prone as people are tend to use `?` or `return` when they get an `Err`.
See what they do in `futures`. `Err` means some error that should be taken care of.
Here you should have some `enum` or `(StorageIterator, bool)` in `Ok` part.